v2/internal/controllers/crd_machinelearningservices_test.go

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package controllers_test
+
+import (
+	"testing"
+
+	"github.com/Azure/azure-service-operator/v2/api/compute/v1beta20201201"
+	"github.com/Azure/azure-service-operator/v2/api/keyvault/v1beta20210401preview"
+	machinelearningservices "github.com/Azure/azure-service-operator/v2/api/machinelearningservices/v1beta20210701"
+	network "github.com/Azure/azure-service-operator/v2/api/network/v1beta20201101"
+	resources "github.com/Azure/azure-service-operator/v2/api/resources/v1beta20200601"
+	storage "github.com/Azure/azure-service-operator/v2/api/storage/v1beta20210401"
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/go-autorest/autorest/to"
+	v1 "k8s.io/api/core/v1"
+)
+
+func Test_MLS_Workspaces_CRUD(t *testing.T) {
+	t.Parallel()
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+
+	sa := newStorageAccount(tc, rg)
+
+	tc.CreateResourceAndWait(sa)
+
+	kv := newVault(tc, rg)
+	tc.CreateResourceAndWait(kv)
+
+	workspaces := newWorkspaces(tc, testcommon.AsOwner(rg), sa, kv)
+
+	tc.CreateResourcesAndWait(workspaces)
+
+	tc.RunParallelSubtests(
+
+		testcommon.Subtest{
+			Name: "Test_WorkspacesCompute_CRUD",
+			Test: func(testContext *testcommon.KubePerTestContext) {
+				WorkspaceCompute_CRUD(tc, testcommon.AsOwner(workspaces), rg)
+			},
+		},
+		testcommon.Subtest{
+			Name: "Test_WorkspacesConnection_CRUD",
+			Test: func(testContext *testcommon.KubePerTestContext) {
+				workspaceConnectionTest(tc, testcommon.AsOwner(workspaces))
+			},
+		},
+	)
+
+	tc.DeleteResourcesAndWait(workspaces, kv, sa, rg)
+
+}
+
+func newWorkspaces(tc *testcommon.KubePerTestContext, owner *genruntime.KnownResourceReference, sa *storage.StorageAccount, kv *v1beta20210401preview.Vault) *machinelearningservices.Workspace {
+	identityType := machinelearningservices.IdentityTypeSystemAssigned
+
+	workspaces := &machinelearningservices.Workspace{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.NoSpaceNamer.GenerateName("work")),
+		Spec: machinelearningservices.Workspaces_Spec{
+			Location: tc.AzureRegion,
+			Owner:    owner,
+			Sku: &machinelearningservices.Sku{
+				Name: to.StringPtr("Standard_S1"),
+				Tier: to.StringPtr("Basic"),
+			},
+			AllowPublicAccessWhenBehindVnet: to.BoolPtr(false),
+			Identity: &machinelearningservices.Identity{
+				Type: &identityType,
+			},
+			StorageAccountReference: tc.MakeReferenceFromResource(sa),
+			KeyVaultReference:       tc.MakeReferenceFromResource(kv),
+		},
+	}
+	return workspaces
+}
+
+func workspaceConnectionTest(tc *testcommon.KubePerTestContext, owner *genruntime.KnownResourceReference) {
+
+	jsonValue := "{\"user\":\"admin\", \"password\":\"abctest\"}"
```

Comment:
There's a `tc.namer` helper to generate passwords, use that for the password part of this.

Also... the fact we're specifying raw JSON seems sorta horrible. Is this an artifact of us doing a bad job generating types or is this really the API?

Reply:
Sorry for the confusion here.. its the value field and should not be password.. will replace with something better like `foo` and `bar`